File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
class Solution {
private:
    TreeNode* getTree(vector<int>& postorder, int postStart, int postEnd, vector<int>& inorder, int inStart, int inEnd, unordered_map<int,int> &inMap){
        if(postStart < postEnd || inStart > inEnd) return NULL;
        
        TreeNode* root = new TreeNode(postorder[postStart]);

        int inRoot = inMap[root->val];
        int numsRight = inEnd - inRoot;

        root->right = getTree(postorder, postStart - 1, postStart - numsRight, inorder, inRoot+1, inEnd, inMap);

        root->left = getTree(postorder, postStart - numsRight - 1, postEnd,
        inorder, inStart, inRoot - 1, inMap);
        
        return root;
        
    }
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>&  postorder) {
        unordered_map<int,int> inMap;
        for(int idx=0; idx<inorder.size(); idx++){
            inMap[inorder[idx]] = idx;
        }
        TreeNode* root = getTree(postorder, postorder.size() - 1,0, inorder, 0, inorder.size()-1, inMap);
        return root;
    }
};

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        inorderLoc = {}

        for i in range(len(inorder)):
            inorderLoc[inorder[i]] = i
        
        postorder = collections.deque(postorder)

        def build(start,end):
            if start > end:
                return None

            root = TreeNode(postorder.pop())

            mid = inorderLoc[root.val]

            root.right = build(mid+1,end)
            root.left = build(start,mid-1)

            return root

        return build(0, len(postorder)-1)